File: genetic-packer/src/main/java/genetic/packer/adapter/RequestResponseAdapter.java
package genetic.packer.adapter;

import java.util.function.BiFunction;
import java.util.function.Function;

import genetic.api.individual.Cell;
import genetic.api.individual.impl.DetailedIndividual;
import genetic.packer.Packer;
import genetic.packer.PackerContextBuilder;
import genetic.packer.dto.request.ContainerDto;
import genetic.packer.dto.request.ParamsDto;
import genetic.packer.dto.request.RequestDto;
import genetic.packer.dto.response.IndividualDto;
import genetic.packer.dto.response.IndividualDtoBuilder;
import genetic.packer.dto.response.ResponseDto;
import genetic.packer.dto.response.ResponseDtoBuilder;
import genetic.packer.evolution.generation.dto.EmbryoBuilder;
import genetic.packer.mapper.BoxMapper;
import javafx.geometry.Bounds;
import javafx.scene.shape.Box;
import javaslang.collection.Seq;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

/**
 * @author piotr.larysz
 */
@Component
public class RequestResponseAdapter implements BiFunction<RequestDto, ParamsDto, ResponseDto> {

    @Autowired
    private Packer packer;

    @Autowired
    private Function<ContainerDto, Bounds> containerToBoundsMapper;

    @Autowired
    private BoxMapper boxMapper;

    @Override
    public ResponseDto apply(RequestDto requestDto, ParamsDto params) {

        final Packer.Context context = new PackerContextBuilder()
                .numberOfGenerations(requestDto.getNumberOfGenerations())
                .generationSize(requestDto.getGenerationSize())
                .numberOfTopIndividuals(requestDto.getNumberOfTopIndividuals())
                .embryo(new EmbryoBuilder()
                    .bounds(containerToBoundsMapper.apply(requestDto.getEmbryo().getContainer()))
                    .boxes(requestDto.getEmbryo().getBoxes().map(boxMapper::map))
                    .build())
                .build();

        final Packer.Result<Double, Box> result = packer.apply(context);

        return new ResponseDtoBuilder()
                .topIndividuals(this.createTopIndividualsDtos(result.getTopIndividuals()))
                .container(requestDto.getEmbryo().getContainer())
                .generationStats(result.getGenerationStats())
                .build();
    }

    public Seq<IndividualDto> createTopIndividualsDtos(Seq<DetailedIndividual<Double, Box>> topIndividuals) {
        return topIndividuals.map(this::createTopIndividualDto);

    }

    public IndividualDto createTopIndividualDto(DetailedIndividual<Double, Box> detailedIndividual) {
        return new IndividualDtoBuilder()
            .fitness(detailedIndividual.getFitness())
            .translatedBoxes(detailedIndividual
                .getCells()
                .map(Cell::getNucleus)
                .map(boxMapper::map)
            )
            .numberOfGeneration(detailedIndividual.getNumberOfGeneration())
            .build();
    }
}

File: genetic-web/src/main/java/genetic/web/adapter/RequestResponseAdapter.kt
package genetic.web.adapter

import genetic.Evolution
import genetic.api.individual.impl.DetailedIndividual
import genetic.packer.dto.request.ContainerDto
import genetic.packer.dto.request.EmbryoDto
import genetic.packer.dto.request.ParamsDto
import genetic.packer.dto.request.RequestDto
import genetic.packer.dto.response.IndividualDto
import genetic.packer.dto.response.ResponseDto
import genetic.packer.evolution.generation.dto.Embryo
import genetic.packer.mapper.BoxMapper
import javafx.geometry.Bounds
import javafx.scene.shape.Box
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Component

@Component
class RequestResponseAdapter(
    private val evolution: Evolution<Box, Embryo>,
    @Qualifier("containerToBoundsMapperImpl") private val containerToBoundsMapper: (ContainerDto) -> Bounds,
    private val boxMapper: BoxMapper
) : (RequestDto<EmbryoDto>, ParamsDto) -> ResponseDto {

  override fun invoke(requestDto: RequestDto<EmbryoDto>, params: ParamsDto) = Evolution.Context(
      numberOfGenerations = requestDto.numberOfGenerations,
      generationSize = requestDto.generationSize,
      numberOfTopIndividuals = requestDto.numberOfTopIndividuals,
      numberOfEliteIndividuals = requestDto.numberOfEliteIndividuals,
      embryo = Embryo(
          bounds = containerToBoundsMapper(requestDto.embryo.container),
          boxes = requestDto.embryo.boxes.map(boxMapper::map)
      )
  )
      .let(evolution)
      .let {
        ResponseDto(
            topIndividuals = it.topIndividuals.map {
              createTopIndividualDto(it)
            },
            container = requestDto.embryo.container,
            generationStats = it.generationStats
        )
      }

  private fun createTopIndividualDto(detailedIndividual: DetailedIndividual<Box>) = IndividualDto(
      fitness = detailedIndividual.fitness.score,
      translatedBoxes = detailedIndividual.cells.map { it.nucleus }
          .map(boxMapper::map),
      numberOfGeneration = detailedIndividual.numberOfGeneration
  )
}